tcmur_cmd_handler.c

```diff
@@ -294,6 +298,657 @@ static int handle_write_verify(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+			     tdll);
+		ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					  ASC_PARAMETER_LIST_LENGTH_ERROR,
+					  NULL);
+		goto err;
+	}
+
+	/*
+	 * From spc4r31, section 6.3.7.1 Segment descriptors introduction
+	 *
+	 * Segment descriptors (see table 120) begin with an eight byte header.
+	 */
+	sdll = be32toh(*(uint32_t *)&par[8]);
+	if (sdll < 8) {
+		tcmu_dev_err(dev, "Illegal segment descriptor length %u\n",
+			     tdll);
+		ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					  ASC_PARAMETER_LIST_LENGTH_ERROR,
+					  NULL);
+		goto err;
+	}
+
+	/*
+	 * The maximum length of the target and segment descriptors permitted
+	 * within a parameter list is indicated by the MAXIMUM DESCRIPTOR LIST
+	 * LENGTH field in the copy managers operating parameters.
+	 */
+	if (tdll + sdll > RCR_OP_MAX_DESC_LIST_LEN) {
+		tcmu_dev_err(dev, "descriptor list length %u exceeds maximum %u\n",
+			     tdll + sdll, RCR_OP_MAX_DESC_LIST_LEN);
+		ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					  ASC_PARAMETER_LIST_LENGTH_ERROR,
+					  NULL);
+		goto err;
+	}
+
+	/*
+	 * The INLINE DATA LENGTH field contains the number of bytes of inline
+	 * data, after the last segment descriptor.
+	 * */
+	inline_dl = be32toh(*(uint32_t *)&par[12]);
+
+	/* From spc4r31, section 6.3.1 EXTENDED COPY command introduction
+	 *
+	 * The EXTENDED COPY parameter list (see table 104) begins with a 16
+	 * byte header.
+	 *
+	 * The data length in CDB should be equal to tdll + sdll + inline_dl
+	 * + parameter list header length
+	 */
+	if (data_length < (XCOPY_HDR_LEN + tdll + sdll + inline_dl)) {
+		tcmu_dev_err(dev, "Illegal list length: length from CDB is %u,"
+			     " but here the length is %u\n",
+			     data_length, tdll + sdll + inline_dl);
+		ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					  ASC_PARAMETER_LIST_LENGTH_ERROR,
+					  NULL);
+		goto err;
+	}
+
+	tcmu_dev_dbg(dev, "Processing XCOPY with tdll: %hu sdll: %u inline_dl: %u\n",
+		     tdll, sdll, inline_dl);
+
+	/*
+	 * Parse the segment descripters and for now we only support block
+	 * -> block type.
+	 *
+	 * The max seg_desc number support is 1(see RCR_OP_MAX_SG_DESC_COUNT)
+	 */
+	seg_desc = par + XCOPY_HDR_LEN + tdll;
+	ret = xcopy_parse_segment_descs(seg_desc, xcopy, sdll, sense);
+	if (ret != SAM_STAT_GOOD)
+		goto err;
+
+	/*
+	 * Parse the target descripter
+	 *
+	 * The max seg_desc number support is 2(see RCR_OP_MAX_TARGET_DESC_COUNT)
+	 */
+	tgt_desc = par + XCOPY_HDR_LEN;
+	ret = xcopy_parse_target_descs(dev, xcopy, tgt_desc, tdll, sense);
+	if (ret != SAM_STAT_GOOD)
+		goto err;
+
+	if (tcmu_get_dev_block_size(xcopy->src_dev) !=
+	    tcmu_get_dev_block_size(xcopy->dst_dev)) {
+		tcmu_dev_err(dev, "The block size of src dev %u != dst dev %u\n",
+			     tcmu_get_dev_block_size(xcopy->src_dev),
+			     tcmu_get_dev_block_size(xcopy->dst_dev));
+		ret = tcmu_set_sense_data(sense, NOT_READY,
+					  ASC_LOGICAL_UNIT_COMMUNICATION_FAILURE,
```

Comment:
You are right, I just missed about che disk block length check.
For the failure the spec doesn't suggest about this, I think COPY ABORTED - INCORRECT COPY TARGET DEVICE TYPE  the  looks better.

Yes, feel free to do the fix :)